Support for Twitter processing

ewan-klein opened the pull request:
This PR provides a basis on which Twitter processing can be built, and consists primarily of:
- a client module, plus demo file etc, in the `twitter`package; the client is built on top of Twython, 
  which is a new external dependency;
- a 'casual text' tokeniser by Chris Potts;
- a corpus reader.

At present, there is no support for Tweet normalisation, POS tagging, or any other downstream linguistic analysis.

In general, it's not possible to carry out automatic regression tests against either the live Twitter stream or past Tweets. However the `nltk/test/unit` directory contains tests for extracting fields from a fixed corpus and tests for the code that reads authentication information from a user file.

The demos will only run if you have set up Twitter authentication for yourself following the instructions in `nltk/test/twitter.ipynb`, which should be run via `ipython notebook` from the command line. The notebook was created under IPython 3.x. More work still needs to be done on fleshing out the examples in this notebook.

Commit:
minor edits

Commit:
added material on filtering the stream

Commit:
Check whether a corpus file is empty

Commit:
examples with userids

Commit:
created twitter package

Commit:
Updates for change in API, SB's comments

Commit:
Added new hydrate() method which writes to file.

Commit:
Changed docstrings
Renamed jsonlist()

Commit:
fixed broken tests, updated for changed API

Commit:
First attempt to define a higher-level Twitter class

Commit:
First import

Commit:
Added more error checking for authentication.
Moved TweetHandler into api.py

Commit:
Renamed interface class

Commit:
Renamed interface class

Commit:
moved utility functions to util.py

Commit:
Added imports to __init__

Commit:
added some imports

Commit:
Moved utility functions into util.py

Commit:
Changed header

Commit:
temporary unittest file in twitter package

Commit:
fixes for tests

Commit:
fixes for tests

Commit:
post refactoring tweaks; changes to catching ImportError

Commit:
changes to catching ImportError

Commit:
changes to catching ImportError

Commit:
changes to catching ImportError

Commit:
changes to catching ImportError

Commit:
Improved error checking

Commit:
minor edits

Commit:
renaming to avoid name clash

Commit:
Mainly PEP8

Commit:
Renaming twitter tokenizer module

Commit:
Renaming corpus readers for Twitter

Commit:
Renaming corpus reader for Twitter

Commit:
Removed boilerplate

Commit:
renaming corpus methods as per #728

Commit:
File writing method for Twitter class.

Commit:
temporary convenience function

Commit:
Merge branch 'develop' into twitter

Commit:
Merge branch 'develop' into twitter

Commit:
temporary file for Twitter documentation

Commit:
fixed bug in Streamer params

Commit:
adding some temporary files for testing

Commit:
documentation files

Commit:
adding some temporary files for testing

Commit:
merged corpus reader info

Commit:
merge with develop branch

Commit:
Moved demo code to separate file.

Commit:
Revert 213430568e4509dac6198acf1190b7949a18cf5d

Commit:
Removed tofile() method from Twitter class

Commit:
Removed 'handler' from signature of Streamer. Added method to get user info from userid.

Commit:
Bugfixes, expanding demo examples.

Commit:
Renamed 'dehydrate', bugfix.

Commit:
minor edits

Commit:
Mainly added more docstrings

Commit:
More docstrings, fixed a few bugs, reorganised order of demos.

Commit:
Fixed various PEP8 errors and warnings

Commit:
issues with Twittter() demo...

Commit:
Added pointer to more emoticons

Commit:
Version of util.py with both field writers

Commit:
Concise version of extract_fields(). Removed json2txt().

Commit:
Updated extract_id demo

Commit:
Updated documentation

Commit:
first chunk of examples

Commit:
first import

Commit:
added new image file and minor updates to notebook

Commit:
Merge branch 'develop' into twitter

Commit:
renaming notebook

Commit:
Updated copyright year

Commit:
moved twitter iPython notebook to nltk/test

Commit:
moved twitter iPython notebook images to nltk/test

Commit:
Merge branch 'develop' into twitter

Commit:
moved location of twitter notebook

Commit:
Backwards compatibility for using file parameter in print function

Commit:
Backwards compatibility for using file parameter in print function

Commit:
Backwards compatibility for html entities

Commit:
Backwards compatibility for html entities

Commit:
PEP8 corrections

Commit:
Remove funny chars from comments

Commit:
* Minor edits to test function
* Moved coding statement to beginning of file

lrnzcig committed:
compatibility python 2.7

super() is not avail for python 2.7

Commit:
added user tweet search

Commit:
added twython to requirements

Commit:
removed old text from doctest file

Commit:
moved oauth test file to unit test dir

Commit:
added a user_tweets method

lrnzcig committed:
compatibility python 2.7

when writing to csv file, in order to provide encoding
- added a new “UnicodeWriter” as csv writer for 2.7 at compat.py
- opening the output file allowing that encoding errors are either
replaced (default), ignored, or throw exception, as in encoding defaults
tested for python 3.4 & 2.7

ewan-klein committed:
Merge pull request #946 from lrnzcig/twitter

compatibility python 2.7

Commit:
Wrapping streaming calls to address dropping connection

lrnzcig committed:
minor corrections to streaming dropping connection

minor corrections:
- capturing only ChunkedEncodingError exception
- formatting the error
- follow parameter

ewan-klein committed:
Merge pull request #950 from lrnzcig/twitter

minor corrections to streaming dropping connection

Commit:
temporary home for tweet sample file

Commit:
Made parameter values to stream.filter() optional

lrnzcig committed:
date limit in streaming requests

initial implementation of parameter for streaming requests that stops
when a date of the tweets stream is reached, regardless of the tweet
count

To be considered/enhanced:
- date is now a python date parameter, should probably accept a string
instead, what format?
- limit in number of counts is still mandatory, consider if it is
necessary

lrnzcig committed:
code readability

as per comments to pull request 951

ewan-klein committed:
Merge pull request #951 from lrnzcig/twitter

date limit in streaming requests

Commit:
added twitter corpus tests

Commit:
Merge branch 'twitter' of https://github.com/nltk/nltk into twitter

Commit:
Merge branch 'develop' into twitter

Commit:
Modified date_limit to use datetime library in place of pytz and modified date checking
in handle() method for TweetWriter. Added a demo for checking date_limit.

Commit:
added twitter corpus reader to corpus.doctest

Commit:
Modifications for Python 2.6 compatibility

Commit:
initial commit of vader code

lrnzcig committed:
compatibility with python 2.7

use pytz to obtain UTC in python <> 3, standard datetime in python 3

lrnzcig committed:
date_limit: dealing with local timezone

“simple” solution for allowing the user to create the date_limit in
their local timezone, i.e. actually not worrying about timezone

avoids use of non-standard package pytz

ewan-klein committed:
Merge pull request #956 from lrnzcig/twitter

Thanks @lrnzcig

lrnzcig committed:
search_tweets with pagination & rate limits

Handles requests to the rest api for count bigger than 100 tweets:
- successive requests using pagination
- wait for 15 minutes when rate limit is reached

lrnzcig committed:
Merge remote-tracking branch 'nltk/twitter' into twitter

ewan-klein committed:
Merge pull request #957 from lrnzcig/twitter

search_tweets with pagination & rate limits

lrnzcig committed:
compatibility with python 2.7

Supposedly tested for pull request #946 (by me) but something went
wrong.

Tested again for python 3.4 & 2.7, encodings cp437, latin1 and utf8
(default)

Commit:
renaming, NLTK boilerplate

lrnzcig committed:
compatibility with python 2.7

write csv files with encoding: testing with field types different to
string

lrnzcig committed:
enhance json2csv fields of other Twitter entities

Enhancement of json2csv function to allow an output csv composed of
fields of different twitter entities

lrnzcig committed:
enhance json2csv fields of other Twitter entities

comments missing

ewan-klein committed:
Merge pull request #958 from lrnzcig/twitter

enhance json2csv fields of other Twitter entities + compatibility 2.7

Commit:
applied 2to3

Commit:
Minor edits. Changed tabs to spaces.

Commit:
lexicon converted to utf-8

lrnzcig committed:
function to csv of entities of a tweet (e.g. hashtag)

new function to extract to csv additional information of the tweet,
what Twitter calles entities: hashtags, usermentions, media & urls

Commit:
Moved demo code to initial doctest block.

ewan-klein committed:
Merge pull request #962 from lrnzcig/twitter

new function to convert to csv add. entities of a tweet (e.g. hashtag)

Commit:
Added new class, plus some refactoring. Removed camelcase names.

Commit:
Some PEP8 fixes

Commit:
Fixed global rename error (s -> sentiment_dict)

Commit:
Prep work on refactoring text preprocessing

Commit:
Extracted text preprocessing into separate method

Commit:
Extracted method for rescoring valence

Commit:
Extracted method to check for presence of 'but'

Commit:
changed n_scalar to module level constant

Commit:
Extracted sentiment_valence method

Commit:
Added SentiText class, and refactored

lrnzcig committed:
new function to convert to csv additional entities

tested for the possible entities of a tweet (hashtags, user_mentions,
urls, media), of a user (urls), and for place (including bounding box)

Commit:
typo

Commit:
updates to twitter corpus handling

Commit:
Merge branch 'twitter' of github.com:nltk/nltk into twitter

Commit:
add check for twython import

Commit:
merge from develop branch

lrnzcig committed:
Merge remote-tracking branch 'nltk/twitter' into twitter

Commit:
updates to tweets corpus reading

lrnzcig committed:
Merge remote-tracking branch 'nltk/twitter' into twitter

lrnzcig committed:
add date limit to rest api

Added date limit functionality to tweets method for rest api (i.e. not
streaming). Api of the method does not change. The param date_limit has
different meaning when either streaming or not; in streaming is a date
in the future, if not is a date in the past

Commit:
minor edits to corpus reader demo

ewan-klein committed:
Merge pull request #968 from lrnzcig/twitter

convert twitter additional entities to csv + date_limit for not streaming accesses

Commit:
Merge branch 'twitter' into sentiment

Commit:
removed explicit numbering from demos

Commit:
some renaming of classes, methods

Commit:
cleaning up

Commit:
Merge branch 'twitter' into sentiment

lrnzcig committed:
add date limit to rest api - bug fix

The new feature to add date limit to rest api was falling into an
infinite loop.

Tested thoughtfully.

Commit:
Minor changes to docstrings. Re-worked error handling.

Commit:
Modified import statement

ewan-klein committed:
Merge pull request #969 from lrnzcig/twitter

add date limit to rest api - bug fix

Commit:
updated notebook

Commit:
Merge branch 'twitter' of github.com:nltk/nltk into twitter

Commit:
Merge branch 'sentiment' into twitter

Commit:
Added case-option for wrapper casual_tokenize function

Commit:
PEP 8 fixes. Updated lexicon reader to use utf-8.

Commit:
Updated id of corpus for TwitterCorpusReader

Commit:
Don't choke on a ZipFilePathPointer.
(Check that this still works for user-installed corpora.)

Commit:
updated import statements from twitter.util

lrnzcig committed:
fix bug tweets function rest api (not streaming)

After some more testing, yet another bug

lrnzcig committed:
python 2.x compatibility

raise from construct is only valid in python 3

lrnzcig committed:
automated test of json2csv function

initial version for review

Commit:
merge + fixed merge conflict

Commit:
Merge branch 'develop' into twitter

Commit:
Don't pass parameter from tweets() method to Tweetviewer init.

Updated docstrings

Commit:
Modified Query's search_tweets() method to return results['statuses']

ewan-klein committed:
Merge pull request #973 from lrnzcig/twitter

automated test of json2csv function

Commit:
Use tweets from corpus rather than personal directory.

Commit:
Partial fix for global variables in demo functions

lrnzcig committed:
json2csv: keep ordering of fields and treat dictionaries

Keep ordering of fields as in the call to the method.

Accept dictionaries as fields.

lrnzcig committed:
automated test of json2csv function

full version with 11 automated tests

lrnzcig committed:
documentation on json2csv function

Initial version of the documentation in a python notebook.

lrnzcig committed:
typo

lrnzcig committed:
locations for streaming, retries for fatal error

minor enhancements:
- added a parameter for streaming using locations as a parameter
- added a retry mechanism for search_tweets (not streaming); if set,
number of retries when searching tweets before raising an exception. By
default 0 retries

ewan-klein committed:
Merge pull request #980 from lrnzcig/twitter

automated test of json2csv function

Commit:
merge from develop branch

Commit:
Reduce overlong lines.

Commit:
PEP8 refactoring

Commit:
* Use TemporaryDirectory for writing files rather than user filespace.
* PEP8 line-wrapping.
* Pass a file-like object rather than filename to json2csv

Commit:
* Updated some docstrings
* Commented out json2csv examples in demo file
* Changed signature of json2csv to take file-like object rather than filename for input
* Explicitly close the output file in json2csv

Commit:
PEP8 line wrapping

lrnzcig committed:
issues with automatic test json2csv

- carriage return depends on machine OS
- made a faster test with a smaller input file (actually first 100
lines of the corpus)
- python 2.7 compatibility of TemporaryDirectory()

ewan-klein committed:
Merge pull request #1010 from lrnzcig/twitter

issues with automatic test json2csv

Commit:
Minor fixes to get Py3 compatibility

Commit:
removing sentiment package from twitter branch

Commit:
removed old twitter docs from branch

Commit:
updates to twitter notebook

lrnzcig committed:
test json2csv adding info for debugging

lrnzcig committed:
test json2csv sort files before compare

sort both output and reference file before compare. Different machines
may provide output in different orders, but still valid

ewan-klein committed:
Merge pull request #1013 from lrnzcig/twitter

test json2csv adding info for debugging

lrnzcig committed:
json2csv_entities bug fix

- making sure that the function gets the info from the main object,
e.g. from the tweet metadata instead of the retweeted metadata
- a couple of lines of the reference files were actually wrong

ewan-klein committed:
Merge pull request #1016 from lrnzcig/twitter

json2csv_entities bug fix

Commit:
Pep8 updates

Commit:
Refactored the test cases. Moved reference files to a subdirectory.

Commit:
merge with develop

Commit:
tidying up twitter branch

Commit:
remove trailing whitespace

Commit:
Expanded instructions on using your own Twitter corpus

Commit:
Updated discussion of twitter corpus

Commit:
removed 1k_sample of Tweets

Commit:
Refactoring Twitter authentication into a class for easier testing.

Commit:
Uodated tests for reading Twitter API Oauth1 credentials from file

Commit:
Changed subdir assignment to class attribute

Commit:
Removed tweetids demos. Tidying up stale code.

Commit:
Renamed twitter test file

Commit:
Updated docstrings, tidying

Commit:
Minor updates to handling authentication

Commit:
Another test file for oauth

Commit:
Removed empty lines

ewan-klein commented:
@stevenbird I'd really like to move forward with this PR. Will you have time to review it by the end of the week?

kmike commented on nltk/tokenize/casual.py:
```diff
@@ -0,0 +1,304 @@
+        """
+        # Try to ensure unicode:
+        try:
+            text = str(text)
```
`text = str(text)` is not correct: if unicode text is passed, in Python 2.x it will try to _encode_ it to bytes using `sys.getdefaultencoding()`.

I think tokenize function should just require unicode input, and the whole "Try to ensure unicode" block should be removed.

ewan-klein replied:
Thanks @kmike. Fixed in https://github.com/nltk/nltk/commit/b15d50a1a50939dba8d7802bd6140e1f47b8a695

kmike commented on nltk/tokenize/casual.py:
```diff
@@ -0,0 +1,304 @@
+    s4 = "RT @facugambande: Ya por arrancar a grabar !!! #TirenTirenTiren vamoo !!"
+    s5 = "http://t.co/7r8d5bVKyA http://t.co/hZpwZe1uKt\
+    http://t.co/ZKb7GKWocy Ничто так не сближает\
+    людей"
```
Note that in Python 2.x all strings are utf8-encoded bytestrings here, so the example doesn't look correct - users should pass unicode to the library. To fix it we can add `from __future__ import unicode_literals` to the top of the file.

kmike commented on nltk/twitter/util.py:
```diff
@@ -0,0 +1,362 @@
+        if isinstance(field, str):
+            out += [json[field]]
+        else:
+            out += [json[value] for value in field]
```
In Python 2.x for unicode strings the second code path will be active. It is better to use `isinstance(field, compat.string_types)`.

kmike commented on nltk/compat.py:
```diff
@@ -121,6 +126,131 @@ def load_module(self, name):
 
     sys.meta_path.insert(0, TkinterLoader())
 
+    import pytz
```
`pytz` is not listed as an NLTK dependency.

ewan-klein replied:
@kmike: well noted. do you think this extra dependency is a problem? 
@lrnzcig: we definitely need the `pytz` import, right?

kmike replied:
I think that pytz dependency is fine, it is a commonly used package. Though it is used only for pytz.utc; there is an example UTC tzinfo implementation in Python 2.x docs: https://docs.python.org/2/library/datetime.html#tzinfo-objects 

```
from datetime import tzinfo, timedelta, datetime

ZERO = timedelta(0)
HOUR = timedelta(hours=1)

# A UTC class.

class UTC(tzinfo):
    """UTC"""

    def utcoffset(self, dt):
        return ZERO

    def tzname(self, dt):
        return "UTC"

    def dst(self, dt):
        return ZERO

utc = UTC()
```

It may be short enough to bundle, without adding a dependency.

kmike commented on nltk/twitter/util.py:
```diff
@@ -0,0 +1,362 @@
+
+
+def _write_to_file(object_fields, items, entity_fields, writer):
+    if items == None:
```
PEP8 "Programming Recommendations" says:

> Comparisons to singletons like None should always be done with is or is not , never the equality operators. 

There is a couple more `== None` comparisons in this PR.

kmike commented on nltk/test/unit/test_json2csv_corpus.py:
```diff
@@ -0,0 +1,184 @@
+    from itertools import izip
+except ImportError:
+        # Python 3
+    izip = zip
```
it is available as nltk.compat.izip

kmike commented on nltk/tokenize/casual.py:
```diff
@@ -0,0 +1,304 @@
+                try:
+                    entnum = int(entnum)
+                    text = text.replace(ent, chr(entnum))
+                except:
```
According to PEP8 ("Programming Recomendations") it is better not to use bare `except:` without specifying an exception type.

ewan-klein replied:
Yes, I noticed this. It's not my code. @kmike: any idea what exception would be thrown in this case?

kmike replied:
It'd be ValueError for int(..) and chr(..) calls, but if the idea is to ignore all errors we can use `except Exception`.

I wonder if it is better to copy-paste an existing battle-tested function (like https://github.com/scrapy/w3lib/blob/606434d6ece995132b4b461b7558687ab0018f8a/w3lib/html.py#L38) or use routines from the standard library (http://stackoverflow.com/a/2087433/114795).

ewan-klein replied:
I think `except Exception` goes against the spirit of the PEP8 recommendation!

@kmike I don't have strong feelings about how to address this; both your suggestions seem reasonable. Given that this seems to be an area where you have experience, would you consider fixing up `casual.py` along the lines you recommend?

kmike replied:
I can send a PR this weekend, is it ok?

ewan-klein replied:
I'd suggest you just push your changes for `casual.py` directly to the twitter branch.

lrnzcig replied:
@ewan-klein done in #1028

Commit:
Removed check for unicode input to tokenizer

ewan-klein commented:
@kmike: Thanks for all the helpful input!

@lrnzcig Most of the revisions proposed so far by @kmike seem easy to implement. I need to stop work now, and will be pretty busy for the next few days. Would you have time over the weekend to deal with the things he mentions?

stevenbird commented:
@kmike – yes, thanks for your input!

stevenbird commented on nltk/corpus/__init__.py:
```diff
@@ -206,6 +206,8 @@
     para_block_reader=tagged_treebank_para_block_reader, encoding='ascii')
 treebank_raw = LazyCorpusLoader(
     'treebank/raw', PlaintextCorpusReader, r'wsj_.*', encoding='ISO-8859-2')
+tweets = LazyCorpusLoader(
```
I'd prefer this to be called `twitter_samples`; we've tended to use the same name for the corpus reader and the corpus directory.

stevenbird commented on nltk/corpus/reader/__init__.py:
```diff
@@ -93,6 +93,7 @@
 from nltk.corpus.reader.udhr import *
 from nltk.corpus.reader.bnc import *
 from nltk.corpus.reader.sentiwordnet import *
+from nltk.corpus.reader.tweets import *
```
`nltk.corpus.reader.twitter`?

stevenbird commented on nltk/test/corpus.doctest:
```diff
@@ -817,6 +817,33 @@ speaker and sentence identifier for a given speech sample:
     >>> timit.utteranceids(spkrid=timit.spkrid(item)) # doctest: +ELLIPSIS
     ['dr1-fvmh0/sa1', 'dr1-fvmh0/sa2', 'dr1-fvmh0/si1466', ...]
 
+twitter
```
I think this would be `twitter_samples`, the name of the corpus

stevenbird commented on nltk/test/corpus.doctest:
```diff
@@ -817,6 +817,33 @@ speaker and sentence identifier for a given speech sample:
+The default tokenizer for Tweets is specialised for 'casual' text, and
+the `tokenized()` method returns a list of lists of tokens.
+
+    >>> tweets.tokenized('tweets.20150430-223406.json')
```
According to the signature of `tokenized()`, we're expecting the sole parameter to be a list of fileids.

ewan-klein replied:
I don't think this is correct. The `fileids` argument of `tokenized()` eventually gets passed to a `CorpusReader` object, whose constructor says:

```
  :param fileids: Specifies the set of fileids for which paths should
            be returned.  Can be None, for all fileids; a list of
            file identifiers, for a specified set of fileids; or a single
            file identifier, for a single file.  Note that the return
            value is always a list of paths, even if ``fileids`` is a
            single file identifier.
```

Do you want me to change the documentation for the twitter corpus reader class to make this clearer?

stevenbird commented on nltk/twitter/twitter_demo.py:
```diff
@@ -0,0 +1,231 @@
+# For license information, see LICENSE.TXT
+
+"""
+Examples to demo the :py:mod:`twitterclient` code.
```
Over the last couple of years I've been trying to migrate demo code into module docstrings, for better accessibility. Is there any chance of doing that here? It might mean putting some utility functions in util.py, a common module in many NLTK packages.

ewan-klein replied:
My reservations about migrating the contents of this file to a module docstring are:
- the demos don't work well as doctests in the ordinary sense since in most case their output is not predictable;
- the demos that 'follow' users in the live stream can take a few minutes to run, so it's good to have a way of selectively excluding them; it would be more complicated to do this in a docstring context;
- as noted, most of the demos won't work without the user first following the instructions in the `test/twitter.ipynb` notebook for obtaining and installing Twitter API keys, so I'd suggest that it would better to point the user to the notebook as an intro to the functionality of the code;
- from this point of view, the demo file could be removed altogether, although it's possibly a useful resource for some users.

stevenbird replied:
I see. The second best place for such demos with docs would be a howto file in `test/twitter.doctest` to appear at http://www.nltk.org/howto/

ewan-klein replied:
I suggest we add a line to the makefile which would convert the .ipynb file to .html and move it to the `howto` directory. But we should also think about how to publish iPython notebooks as HOWTOs.

ewan-klein replied:
Just to clarify: it's fine as far as I'm concerned to move `twitter_demo.py` to the `/test` directory, but I don't think there would be a lot to gain. For reasons mentioned above, it wouldn't make much sense to convert the file to doctest format (we'd have to skip almost all the outputs), so we wouldn't get useful HTML out of it. And we can't use it as a source of automated testing unless the test harness has Twitter API keys available. Going the IPython notebook route still strikes me as the best option.

stevenbird commented on nltk/twitter/twitterclient.py:
```diff
@@ -0,0 +1,457 @@
+
+
+"""
+NLTK Twitter client
```
This would be a nice discoverable place for examples of usage.

lrnzcig commented:
@ewan-klein No problem, I went through the comments quickly and agreed most of them should be easy. I'll spend some time on it today and on Sunday and I'll send a PR.

@kmike thanks for your comments. I'll try to implement the change you propose for getting rid of `pytz`, indeed it is just to solve the timezone of the tweets. If you want, I can try to take care of `casual.py` as well.

lrnzcig committed:
fixed comments on PR #1021

- compatibility python2.7 + removed dependency with pytz
- naming
- code style

tested in both python 2.7 and 3
- automatic test_json2csv_corpus
- get tweets and streaming
- nltk.download() just in case

lrnzcig committed:
fixed comments on PR

lrnzcig commented:
Ok. I have to leave it here for now. I'll be able to spend some more time on it on Sunday.

I fixed all the comments except the following:
- the `html2unicode` method in `casual.py`, comment by @kmike (I'd like to do it, but just run out of time)
- the `tokenized()` method of twitter corpus, comment by @stevenbird (the change is not difficult but this was done by @ewan-klein and I'd like to know his opinion)
- the comment by @stevenbird about _module docstrings_ (second last comment) -honestly I don't quite understand what it means
- the comment by @stevenbird about the _nice discoverable place for examples of usage_ (last comment) -I don't understand it either

ewan-klein committed:
Merge pull request #1026 from lrnzcig/twitter

fixed comments on PR #1021

stevenbird commented:
@lrnzcig – please see, e.g. tag/regexp.py for an example of a module docstring with demonstration code. Unlike demo modules, these doctests are tested by our CI server, and are accessible via Python's help function.

jnothman commented on nltk/twitter/util.py:
```diff
@@ -0,0 +1,362 @@
+    <https://dev.twitter.com/overview/api/tweets> for a full list of fields.
+    e. g.: ['id_str'], ['id', 'text', 'favorite_count', 'retweet_count']
+    Addionally, it allows IDs from other Twitter objects, e. g.,\
+    ['id', 'text', {'user' : ['id', 'followers_count', 'friends_count']}]
```
Using dict to specify nested objects leaves column order unspecified, which is problematic if headers are not output; and if headers are output the column names should be specified here. (Presumably one can use an `OrderedDict` to solve the first of those problems but who does?) Why not just use a notation like `'user.id', 'user.followers_count'` that keeps the parameter as flat as the output?

lrnzcig replied:
@ewan-klein done in #1028

jnothman commented on nltk/twitter/util.py:
```diff
@@ -0,0 +1,362 @@
+
+    This utility function allows a file of full tweets to be easily converted
+    to a CSV file for easier processing of Twitter entities. For example, the
+    hashtags or media elements of a tweet can be extracted.
```
It's not clear from this description how a one-to-many tweet/entities mapping is to be encoded as CSV. I presume it is one entity per row.

lrnzcig replied:
@ewan-klein done in #1028

jnothman commented on nltk/twitter/util.py:
```diff
@@ -0,0 +1,362 @@
+    of the object that corresponds to the key of the dictionary (could be\
+    the user object, or the place of a tweet object).
+
+    :param list entity_name: The name of the entity: 'hashtags', 'media',\
```
Can we call this `entity_type`, rather than name? Presumably a mentioned user is the entity, and has a name.

lrnzcig replied:
@ewan-klein done in #1028

jnothman commented on nltk/twitter/twitterclient.py:
```diff
@@ -0,0 +1,457 @@
+        return itertools.chain.from_iterable(listoflists)
+
+
+    def lookup(self, infile, outfile, verbose=True):
```
This method doesn't seem to be called anywhere in the PR (at least not with `.lookup(`).

ewan-klein replied:
This was introduced on the assumption that we would only be allowed to distribute Tweet IDs, not full Tweets. As we subsequently received a clarification from Twitter that we can in fact include up to 50k Tweets in the NLTK data distribution, the use case for `lookup()` is much less prominent, and I decided to remove the usage examples from the `twitter_demo.py` file. 

I'm in two minds about what to do now. It might be better packaged up as a standalone command line utility, since it probably won't form part of a Python processing pipeline. If it's a standalone utility, then I'd be inclined to leave it to the user to address compressing / decompressing file inputs and outputs.

jnothman commented on nltk/twitter/twitterclient.py:
```diff
@@ -0,0 +1,457 @@
+        Tweets (if they haven't been deleted) and dump them in a file.
+
+        :param str infile: Name of a file consisting of tweetIDs, one to a line
+        :param str outfile: Name of file where JSON serialisations of fully\
```
It should be clear here -- or is it elsewhere? -- that the output file is _not_ valid JSON, rather one JSON object per line.

jnothman commented on nltk/twitter/twitterclient.py:
```diff
@@ -0,0 +1,457 @@
+        if os.path.isfile(outfile):
+            os.remove(outfile)
+
+        with open(outfile, 'a') as f:
```
In the context of something as voluminous at Twitter data, I am rather wary of methods that only accept _paths_ to files, rather than file-like objects that might incorporate compression. In general, I think it is bad Python design to admit paths exclusively to an I/O interface.

jnothman commented on nltk/twitter/api.py:
```diff
@@ -0,0 +1,86 @@
+    def handle_chunk(self, data_chunk):
+        """
+        Deal appropriately with a list of elements returned by the Twitter API
+        (default implementation should be enough in most cases)
```
What does the return value mean? `handle_chunk` is not, to me, a boolean preposition.

lrnzcig replied:
@ewan-klein done in #1028

jnothman commented on nltk/twitter/twitterclient.py:
```diff
@@ -0,0 +1,457 @@
+            print("""Written {0} Tweets to file {1} of length {2}
+            bytes""".format(count, outfile, os.path.getsize(outfile)))
+
+    def search_tweets(self, keywords, count=100, lang='en', retries_after_twython_exception=0):
```
For a few reasons, this would be nicer as a generator. If one lands up in that 15-minute sleep statement, and decides they'd rather look at what they've got so far, they have the choice of waiting, or killing the operation and failing to re-request it.

It also allows the user to determine `count` dynamically. But keeping the `count` parameter means we don't force users to know `itertools.islice`.

ewan-klein replied:
@lrnzcig: Any thoughts about implementing @jnothman's suggestion?

lrnzcig replied:
I would be happy to figure out how to do it but honestly I don't have much experience with python generators (read about them only after @jnothman's comment.) I'm quite confused by the point of the possibility of setting `count` dynamically -I've been reading about `itertools.islice` and I would not know how to do it, or maybe I'm misunderstanding "dynamically" in this context.

Anyway, I would not go for a dynamic setting of `count`. In my experience when fetching tweets, I've found it was more useful to set a high count as an upper limit in the number of tweets, but using `date_limit` as the way to really stop the process.

Conclusion: I can try the generator part, yes.

jnothman commented on nltk/twitter/twitterclient.py:
```diff
@@ -0,0 +1,457 @@
+        :param int limit: number of data items to process in the current\
+        round of processing
+
+        :param bool repeat: flag to determine whether multiple files should be\
```
Option to compress, or provide a hook (or overridable method) for a custom `open`.

lrnzcig replied:
@ewan-klein added option to compress for `json2csv` functions in PR #1028, using gzip.

For `TweetWriter` I did some trials but faced compatibility problems in 2.7 and 3.4, and anyhow this will probably change if converting into a generator as per previous comment?

jnothman commented on nltk/corpus/reader/twitter.py:
```diff
@@ -0,0 +1,157 @@
+
+        :param word_tokenizer: Tokenizer for breaking the text of Tweets into
+        smaller units, including but not limited to words.
+
```
Many applications may want to filter a stored stream (e.g. by language). Allowing the user to provide a filter function applied before `'text'` alone is extracted may help.

ewan-klein replied:
@jnothman This seems a reasonable idea, but I'd be inclined to keep things simple until someone says they want this extra functionality. But happy for you to add filter as a parameter if you feel that's the right thing to do.

lrnzcig committed:
comments PR #1021

handler api should return no values

lrnzcig committed:
comments PR #1021

handler api should return no values

lrnzcig committed:
comments PR #1021

refactoring json2csv functions to make them more usable: hierarchical
fields instead of dictionaries

lrnzcig committed:
comments PR #1021

adding header to json2csv functions

lrnzcig committed:
adapting notebook to latest changes

- provide input file correctly
- json2csv entries are not dictionaries anymore

lrnzcig committed:
comments PR #1021

substituting function for converting html entities (from w3lib)

lrnzcig committed:
adding compression to output files of json2csv

using gzip, by default the output is compressed

tested in python 2.7 & 3

lrnzcig committed:
adding compression to output files of json2csv

modifying python notebook accordingly

ewan-klein committed:
Merge pull request #1028 from lrnzcig/twitter

Fixes by @lrnzcig to comments on PR #1021

lrnzcig committed:
gzip compression for tweet files

for both streaming and rest

tested for python2.7 and 3

Commit:
Updated corpus name to 'twitter_samples'

Commit:
Minor updates to documentation

lrnzcig committed:
provide output of search_tweets as python generator

tested for 2.7 and 3

lrnzcig committed:
more tests after compression & python generator

ewan-klein committed:
Merge pull request #1034 from lrnzcig/twitter

Twitter compressed output files + generator (comments PR #1021)

Commit:
Clarified output of lookup()

Commit:
Fixed typo

Commit:
Merge branch 'twitter' of github.com:nltk/nltk into twitter

fievelk committed:
add word lengthening reduction function

fievelk committed:
fix comment

fievelk committed:
set default normalization parameter value to True

fievelk committed:
refactor normalization and fix test

fievelk committed:
implement method to strip username handles

fievelk committed:
refactor test method and add normalized examples

fievelk committed:
refactor twitter handles regular expression

Add test examples.

fievelk committed:
edit comments

fievelk committed:
add doctest exaples replacing test function

Commit:
Removed lookup() method, and modified and renamed _lookup() to
expand_tweetids().
Cosmetic changes to docstrings and comments

Commit:
Changed parameter name in user_tweets() from count
to limit for consistency with other methods.

Commit:
Added demo to illustrate expand_tweetids()

lrnzcig committed:
simplify Twitter.tweets() api and repeat

- remove `retries` and `locations` from Twitter.tweets() api, in order
to keep it simple
- recover `repeat`
- by default no compression in json2csv functions

tested for 2.7 and 3

lrnzcig committed:
Twitter.tweets when keywords find no tweets

ewan-klein committed:
Merge pull request #1039 from lrnzcig/twitter

simplify Twitter.tweets() api and restore repeat parameter

Commit:
Merge branch 'develop' into twitter

ewan-klein committed:
Merge pull request #1035 from fievelk/twitter

Twitter tokenizer normalization functions

Commit:
PEP8 fixes, plus tweaks to docstrings.

stevenbird commented:
Thanks @lrnzcig, @ewan-klein, @kmike, @jnothman.

stevenbird commented:
@ewan-klein – note that this PR has given rise to some run-time errors on our continuous integration server at
https://nltk.ci.cloudbees.com/job/nltk/

lrnzcig commented:
@stevenbird @ewan-klein just took a quick look and most of the errors have to do with file locations - some of the added tests look for files at paths relative to the execution path, and don't find them. During the weekend I'll try to take a look at the configuration in Jenkins to figure it out.

BTW a couple of the failing tests seem to have nothing to do with this PR (e.g. gensim_doctest)

stevenbird commented:
@lrnzcig if the tests require data to be installed, then it will need to be added to the NLTK data collection and listed in https://github.com/nltk/nltk_data/blob/gh-pages/collections/all.xml. I'm happy to help with this.

lrnzcig commented:
@stevenbird I'm don't think it is related to NLTK data collections. The link you provide is for adding corpora, if I understood it right.

The tests that are failing use reference files in relative paths, e.g. `nltk.test.unit.test_json2csv_corpus.TestJSON2CSV.test_textoutput` uses a reference file in the relative path `files/tweets.20150430-223406.text.csv.ref`, and the output file is compared to the reference file. The problem in the Jenkins setup is that the test does not find the reference file in the relative path.

Actually, I could reproduce it by running it from my machine shell. If I do
`python ~/git/nltk/nltk/test/unit/test_json2csv_corpus.py`
from the directory `~/git/nltk/nltk/test/unit` everything works fine, but if I do it from any other directory, if fails the same way as Jenkins.

I've modified the way the relative paths are formed in PR #1043, and it works from my shell now. Hopefully it will work in Jenkins as well.
